internal/bufferedread/prefetcher.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufferedread
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/googlecloudplatform/gcsfuse/v3/common"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/block"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/workerpool"
+	"github.com/googlecloudplatform/gcsfuse/v3/metrics"
+)
+
+// prefetcher encapsulates the state and logic for prefetching blocks from GCS. It
+// is responsible for managing the prefetch window, scheduling download tasks,
+// and handling the multiplicative increase in the number of blocks to prefetch.
+type prefetcher struct {
+	object       *gcs.MinObject
+	bucket       gcs.Bucket
+	pool         *block.GenBlockPool[block.PrefetchBlock]
+	workerPool   workerpool.WorkerPool
+	queue        common.Queue[*blockQueueEntry]
+	metricHandle metrics.MetricHandle
+	config       *BufferedReadConfig
+	readerCtx    context.Context
+	readHandle   []byte // For zonal bucket.
+
+	// nextBlockIndexToPrefetch is the index of the next block to be prefetched.
+	nextBlockIndexToPrefetch int64
+
+	// numPrefetchBlocks is the number of blocks to prefetch in the next
+	// prefetching operation.
+	numPrefetchBlocks int64
+
+	// prefetchMultiplier is the factor by which numPrefetchBlocks is multiplied
+	// after a successful prefetch cycle.
+	prefetchMultiplier int64
+}
+
+// prefetcherOptions holds the dependencies for creating a new Prefetcher.
+type prefetcherOptions struct {
```

Comment:
Doesn't seem this refactoring important other than moving the method to different file - Extra overhead of keeping multiple common variable without getting benefit other than moving the implementation to different file.